theano/gpuarray/opt_util.py

```diff
@@ -329,3 +329,48 @@ def opt(node):
             return maker(node, inputs)
         return opt
     return wrapper
+
+
+def pad_dims(input, leftdims, rightdims):
+    """Reshapes the input to a (leftdims + rightdims) tensor
+
+    This helper function is used to convert pooling inputs with arbitrary
+    non-pooling dimensions to the correct number of dimensions for the
+    GPU pooling ops.
+
+    This reduces or expands the number of dimensions of the input to
+    exactly `leftdims`, by adding extra dimensions on the left or by
+    combining some existing dimensions on the left of the input.
+    """
+    assert input.ndim >= rightdims
+
+    if input.ndim == (leftdims + rightdims):
+        return input
+
+    # extract image dimensions
+    img_shape = input.shape[-rightdims:]
+
+    # count the number of "leading" dimensions, store as dmatrix
+    batch_size = tensor.prod(input.shape[:-rightdims])
+    batch_size = tensor.shape_padright(batch_size, 1)
```

Comment:
These lines and the tensor.cast on the next lines come from the original [pool_2d (theano/tensor/signal/pool.py#L95)](https://github.com/Theano/Theano/blob/166cac0/theano/tensor/signal/pool.py#L95-L97).

It seems correct to me (and it doesn't work without this line). The purpose seems to be to compress all leading non-pooling dimensions into the first dimension and setting the second non-pooling dimension to 1.
1. Line 354 computes the product of the length all of the leading non-pooling dimensions.
2. Line 355 converts the scalar batch_size to a vector, which is required by `tensor.join`.
3. Line 359 uses tensor.join to concatenate the vectors `[batch_size], [1], img_shape`.